components/markdown_image_expand/markdown_image_expand.tsx

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import React, {useEffect} from 'react';
+import './markdown_image_expand.scss';
+
+export type Props = {
+    alt: string;
+    imageKey: string;
+    children: React.ReactNode;
+    isExpanded: boolean;
+    postId: string;
+    onToggle: (isExpanded: boolean) => void;
+    actions: {
+        toggleInlineImageVisibility: (postId: string, imageKey: string) => void;
+    };
+};
+
+const MarkdownImageExpand: React.FC<Props> = ({children, alt, isExpanded, postId, actions, onToggle, imageKey}: Props) => {
+    const {toggleInlineImageVisibility} = actions;
+
+    useEffect(() => {
```

Comment:
This `useEffect` block /feels/ wrong -- listening a prop so we can trigger a prop callback wouldn't, ideally, be the responsibility of any component.

Does it suffice to invoke `onToggle` only within `handleToggleButtonClick`? Presumably the "first render" will already have the necessary state for `isExpanded`, allowing the post height computations to occur organically. Thereafter, I'd expect the prop to change only because we toggle the visibility via the click event.

Thoughts?

Reply:
Oh yeah, actually this is a pretty big item. Let me try to explain myself.

So, the thing is that we execute `onToggle` callback we need to provide some data inside. Which is - current toggle state (whether it is expanded or collapsed). 

And a big question here - where should this "computational" logic reside? I mean, logic which should compute next state for our toggle out of the old one? 

Correctly our logic for this state toggling resides within action creators (`toggleInlineImageVisibility` specifically). If I get your approach correctly, in this case we will have to either:

1. Duplicate this logic inside this component as well (do something like `onToggle(!isExpanded)`)
2. Or maybe delegate this logic exclusively to the component and use store just as a storage

Point 2 has come to mind just now to be honest. On the other hand, this kind of "store action isolation inside one component" might be bad - as soon as we wanted to toggle inline image visibility from some other component.

Reply:
The main thing I don't like about this implementation is data flow which looks kinda bi-directional in this case.
Let's assume that user click "Expand" button - here how the flow goes:
1. Action gets dispatched to store
2. Store propagates new `isExpanded` state to a `MarkdownImageExpand`
3. `MarkdownImageExpand` tracks update (via this `useEffect`) and propagates these changes back to `PostMessageView` component

I have been thinking about flattening this flow to something unidirectional. Can we select something from a store that would for sure change when visibility of some inline image changed and pass it as a prop for `PostMessageView`?

I guess we can somehow select all of the storage elements prefixed with `INLINE_IMAGE_VISIBILITY` and... compare previous visibility array with new one? 

I am sorry if I am overthinking this.

Reply:
@lieut-data I think the thing that you're missing with those two props is that one comes from the parent component (`onToggle`) and the other comes from redux (`isExpanded`), so it's not really possible for the parent to be responsible for both.

I'm not a huge fan of how it works right now, but I can't see any way around it unless we can give the parent component some concept of "this is the expanded/collapsed state of my children" like @TheDarkestDay is suggesting, but that does seem harder to represent given how we store the data in local storage. I personally think the current way is our best bet since it's almost equivalent to treating `isExpanded` as if it was part of the state of the `MarkdownImageExpand` except for it being stored in redux.

Reply:
@hmhealey, yeah the source of the props threw me for a loop the first time around, since I didn't see it being passed by the parent and only later realized it was tacked on via `connect`. But my comment about it feeling "wrong" was framed with that perspective already in mind: triggering an event from a prop change to a prop callback seems like an anti-pattern regardless of the source of those props.

**But** I suppose this is really just the hook equivalent of a `componentDidUpdate` lifecycle event that detects the props change. And, for better or worse, we do have this pattern all over the code. So I'm fine leaving as-is :)